### ISSUE TITLE
Snaps "Automatic refreshes" broke Cluster

### ISSUE THREAD

**Issue by lampsolutions** [lxc/lxd#5045]
Snaps feature  "Automatic refreshes" broke my lxd cluster.
Please see attached syslog files from the 3 servers and look for 

```
lxd.daemon ... => Stop reason is: snap refresh
```

Server 1:
```
root@helix01:~# snap list
Name  Version  Rev   Tracking  Publisher   Notes
core  16-2.35  5328  stable    canonical✓  core
lxd   3.0.2    8701  3.0       canonical✓  -
```

Server 2:
```
root@helix02:~# snap list
Name  Version  Rev   Tracking  Publisher   Notes
core  16-2.35  5328  stable    canonical✓  core
lxd   3.0.2    8701  3.0       canonical✓  -
```

Server 3:
```
root@helix03:~# snap list
Name  Version  Rev   Tracking  Publisher   Notes
core  16-2.35  5328  stable    canonical✓  core
lxd   3.0.1    8028  3.0       canonical✓  -
```

At first i tried to reload the lxd daemon on all 3 servers, which worked for server 1 and 2 but not for server 3.
Trying to start lxd on server 3 outputted the following in debug mode:
```
root@helix03:~# sudo lxd --debug --group lxd
DBUG[09-18|11:05:37] Connecting to a local LXD over a Unix socket 
DBUG[09-18|11:05:37] Sending request to LXD                   etag= method=GET url=http://unix.socket/1.0
INFO[09-18|11:05:37] LXD 3.0.1 is starting in normal mode     path=/var/snap/lxd/common/lxd
INFO[09-18|11:05:37] Kernel uid/gid map: 
INFO[09-18|11:05:37]  - u 0 0 4294967295 
INFO[09-18|11:05:37]  - g 0 0 4294967295 
INFO[09-18|11:05:37] Configured LXD uid/gid map: 
INFO[09-18|11:05:37]  - u 0 1000000 1000000000 
INFO[09-18|11:05:37]  - g 0 1000000 1000000000 
WARN[09-18|11:05:37] CGroup memory swap accounting is disabled, swap limits will be ignored. 
INFO[09-18|11:05:37] Initializing local database 
INFO[09-18|11:05:37] Initializing database gateway 
INFO[09-18|11:05:37] Start database node                      address=10.10.10.2:18443 id=3
INFO[09-18|11:05:37] Raft: Restored from snapshot 92-745698-1537268305614 
INFO[09-18|11:05:37] Raft: Initial configuration (index=77): [{Suffrage:Voter ID:1 Address:0} {Suffrage:Voter ID:2 Address:10.10.10.84:18443} {Suffrage:Voter ID:3 Address:10.10.10.2:18443}] 
INFO[09-18|11:05:37] Raft: Node at 10.10.10.2:18443 [Follower] entering Follower state (Leader: "") 
INFO[09-18|11:05:37] LXD isn't socket activated 
DBUG[09-18|11:05:37] Connecting to a local LXD over a Unix socket 
DBUG[09-18|11:05:37] Sending request to LXD                   etag= method=GET url=http://unix.socket/1.0
DBUG[09-18|11:05:37] Detected stale unix socket, deleting 
DBUG[09-18|11:05:37] Detected stale unix socket, deleting 
INFO[09-18|11:05:37] Starting /dev/lxd handler: 
INFO[09-18|11:05:37]  - binding devlxd socket                 socket=/var/snap/lxd/common/lxd/devlxd/sock
INFO[09-18|11:05:37] REST API daemon: 
INFO[09-18|11:05:37]  - binding Unix socket                   socket=/var/snap/lxd/common/lxd/unix.socket
INFO[09-18|11:05:37]  - binding TCP socket                    socket=10.10.10.2:18443
INFO[09-18|11:05:37] Initializing global database 
DBUG[09-18|11:05:37] Failed to establish gRPC connection with 10.10.10.32:18443: 404 Not Found 
DBUG[09-18|11:05:37] Failed to establish gRPC connection with 10.10.10.84:18443: 404 Not Found 
DBUG[09-18|11:05:37] Found cert                               k=0
DBUG[09-18|11:05:37] Failed to establish gRPC connection with 10.10.10.2:18443: 503 Service Unavailable 
DBUG[09-18|11:05:37] Failed to establish gRPC connection with 10.10.10.32:18443: 404 Not Found 
DBUG[09-18|11:05:37] Failed to establish gRPC connection with 10.10.10.84:18443: 404 Not Found 
DBUG[09-18|11:05:37] Found cert                               k=0
DBUG[09-18|11:05:37] Failed to establish gRPC connection with 10.10.10.2:18443: 503 Service Unavailable 
DBUG[09-18|11:05:38] Failed to establish gRPC connection with 10.10.10.32:18443: 404 Not Found 
DBUG[09-18|11:05:38] Failed to establish gRPC connection with 10.10.10.84:18443: 404 Not Found 
DBUG[09-18|11:05:38] Found cert                               k=0
DBUG[09-18|11:05:38] Failed to establish gRPC connection with 10.10.10.2:18443: 503 Service Unavailable 
DBUG[09-18|11:05:38] Failed to establish gRPC connection with 10.10.10.32:18443: 404 Not Found 
DBUG[09-18|11:05:38] Failed to establish gRPC connection with 10.10.10.84:18443: 404 Not Found 
DBUG[09-18|11:05:38] Found cert                               k=0
DBUG[09-18|11:05:38] Failed to establish gRPC connection with 10.10.10.2:18443: 503 Service Unavailable 
DBUG[09-18|11:05:38] Found cert                               k=0
INFO[09-18|11:05:38] Raft-http: Establishing new connection with 10.10.10.32:18443 
DBUG[09-18|11:05:38] Raft-net: 10.10.10.2:18443 accepted connection from: 10.10.10.32:54914 
panic: wal frames txn 745700 (1): out of memory

trace:
2018-09-18 11:05:37.21763: fsm: restore=745698 start
2018-09-18 11:05:37.21763: fsm: restore=745698 database size: 4096
2018-09-18 11:05:37.21764: fsm: restore=745698 wal size: 267832
2018-09-18 11:05:37.21789: fsm: restore=745698 filename: db.bin
2018-09-18 11:05:37.21790: fsm: restore=745698 transaction ID: 
2018-09-18 11:05:37.21846: fsm: restore=745698 open follower: db.bin
2018-09-18 11:05:37.21987: fsm: restore=745698 done
2018-09-18 11:05:38.62932: fsm: term=93 index=745701 cmd=frames txn=745700 pages=0 commit=1 start
2018-09-18 11:05:38.62945: fsm: term=93 index=745701 cmd=frames failed: wal frames txn 745700 (1): out of memory


goroutine 10 [running]:
github.com/CanonicalLtd/dqlite/internal/trace.(*Tracer).Panic(0xc4202c57c0, 0xfff699, 0x2, 0xc42062bd20, 0x1, 0x1)
	/build/lxd/parts/lxd/go/src/github.com/CanonicalLtd/dqlite/internal/trace/tracer.go:59 +0x12c
github.com/CanonicalLtd/dqlite/internal/replication.(*FSM).Apply(0xc420369300, 0xc420218150, 0x0, 0x0)
	/build/lxd/parts/lxd/go/src/github.com/CanonicalLtd/dqlite/internal/replication/fsm.go:84 +0x137
github.com/hashicorp/raft.(*Raft).runFSM.func1(0xc42030f870)
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/fsm.go:57 +0x15a
github.com/hashicorp/raft.(*Raft).runFSM(0xc42022cb00)
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/fsm.go:120 +0x2fa
github.com/hashicorp/raft.(*Raft).(github.com/hashicorp/raft.runFSM)-fm()
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/api.go:506 +0x2a
github.com/hashicorp/raft.(*raftState).goFunc.func1(0xc42022cb00, 0xc4204d63a0)
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/state.go:146 +0x53
created by github.com/hashicorp/raft.(*raftState).goFunc
	/build/lxd/parts/lxd/go/src/github.com/hashicorp/raft/state.go:144 +0x66
```

The problem was fixed by running "snap refresh" on server3.
I also attached syslog from lxd for the three servers.

[syslog-helix01.txt](https://github.com/lxc/lxd/files/2392501/syslog-helix01.txt)
[syslog-helix02.txt](https://github.com/lxc/lxd/files/2392503/syslog-helix02.txt)
[syslog-helix03.txt](https://github.com/lxc/lxd/files/2392505/syslog-helix03.txt)

**Comment by lampsolutions** [comment]
> You need to upgrade Server 3 as well with snap refresh, to match the version of Server 1 and 2.

Hi thanks for your response.
As i already said at the end of my issue ticket i did that and the cluster is running again.

However i think there should be a readable message or something like that which states your version is not compatible and not an critical panic out of memory error message. ;-)
```
panic: wal frames txn 745700 (1): out of memory

**Comment by stgraber** [comment]
Ok, so I think we can actually close this issue as everything mentioned here has actually been fixed.
The problem is that LXD 3.0.1 didn't have any of the fixes so this one upgrade was still a mess...
Things should go much much smoother when going to 3.0.3.

The panic I believe is because of 3.0.1 trying to talk to 3.0.2 where the database was changed to a different protocol and the old implementation didn't have much in the way of guarding for such changes.

The hang during upgrade that happened is also fixed starting with 3.0.2. LXD now has logic to automatically detect when another cluster node has updated to something more recent than itself and will self-initiate a `snap refresh` to get itself onto that version.

We've seen this at work when people starting upgrading from 3.4 to 3.5 and it seems to have been working mostly as intended (only had one report where a user had a stuck cluster member that wouldn't shut down and so was holding the cluster).
